### PR TITLE
feat(backend): implement dynamic keras code generation via registryFeature/ssot dynamic generation

### LINKED ISSUE #177: [Architecture Proposal] Phase 1 Foundation: Dynamic Data-Driven Layer Registry (Single Source of Truth)
## Overview

As outlined in the GSoC 2026 Phase 1 Expected Results, TensorMap needs to transition from a hardcoded layer system (requiring 6+ file edits) to a data-driven layer registry. 

To achieve this and easily scale to 15+ layer types (CNNs, RNNs, Regularization), I propose a **Single Source of Truth (SSOT)** architecture. By defining layers in a central JSON registry served by the backend, we can decouple the frontend UI from the backend generation logic.

## Current Bottleneck (The 6-File Problem)

Currently, adding a layer like `Dropout` requires:

1. Hardcoding the button in `Sidebar.jsx`
2. Creating a dedicated `DropoutNode.jsx` UI component
3. Registering the node in `Canvas.jsx`
4. Hardcoding Pydantic schemas in `deep_learning.py`
5. Updating the code generation engine
6. Updating the Keras compilation mapping

## Proposed Architecture: The Unified Registry

I propose creating a `layer_registry.json` on the backend and serving it via a new endpoint (`GET /api/v1/layers`). Both the FastAPI backend and React frontend will use this single contract.

### 1. The JSON Contract (Example)

```json
{
  "layers": {
    "dense": {
      "display_name": "Dense",
      "category": "core",
      "keras_mapping": "tf.keras.layers.Dense",
      "params": {
        "units": { "type": "number", "default": 32, "min": 1, "required": true },
        "activation": { "type": "select", "options": ["relu", "sigmoid", "softmax", "linear"], "default": "relu" }
      }
    },
    "dropout": {
      "display_name": "Dropout",
      "category": "regularization",
      "keras_mapping": "tf.keras.layers.Dropout",
      "params": {
        "rate": { "type": "number", "default": 0.5, "min": 0.0, "max": 1.0, "required": true }
      }
    }
  }
}
````

### 2. Frontend Implementation (Dynamic Rendering)

**Dynamic Sidebar:**
`Sidebar.jsx` fetches the registry on load and maps through layers to generate the draggable buttons, grouping them by category.

**Universal Node Component:**
Instead of a separate `.jsx` file for every layer, we build a single `GenericLayerNode.jsx`. When dropped on the canvas, it looks up its type in the registry and dynamically maps the `params` object to UI inputs (rendering a `<Select>` for `activation` and a number `<Input>` for `units`).

### 3. Backend Implementation (Dynamic Validation)

**Pydantic Dynamic Models:**
Instead of hardcoded models, we use `create_model` in Pydantic to dynamically generate validation schemas based on the JSON registry at server startup.

**Streamlined Code Gen:**
The generation engine loops through the ReactFlow payload, looks up the `keras_mapping` string, and dynamically unpacks the frontend payload into kwargs:

```python
f"{layer.keras_mapping}(**{layer.params})"
```
## Action Plan & Phased Rollout

If the maintainers are aligned with this architecture, I am ready to lead Phase 1 with the following PRs:

* **PR 1:** Define the `layer_registry.json` and build the `GET /api/v1/layers` endpoint.
* **PR 2:** Build the frontend `GenericLayerNode.jsx` to dynamically consume the API and replace the existing hardcoded Dense/Conv2D nodes.
* **PR 3:** Update backend code generation to use the registry dictionary.

@ivantha I'd love your thoughts on this SSOT approach. Does this align with your vision for the Phase 1 Foundation?

### PR DESCRIPTION
**Pull request by NamanmeetSingh**
## Description

As the final step (Part 3) of the Phase 1 Foundation architecture, this PR completely refactors the Keras model generation logic to be dynamically driven by the Single Source of Truth `layer_registry.json`.

**Issues Addressed:**
* **Scalability Bottleneck:** The backend previously relied on a massive, hardcoded `if/elif` block to translate ReactFlow nodes into TensorFlow code. Adding a new layer required manual backend modifications.
* **Data Loss:** Pydantic validation was silently stripping the `registry` payload sent by the new dynamic frontend.

**Specific Changes Made:**
* **`model_generation.py`**: 
  * Deleted the hardcoded `_build_layer` `if/elif` blocks.
  * Implemented pure dynamic instantiation using `getattr(tf.keras.layers, class_name)` based on the `keras_mapping` provided by the UI.
  * Added robust Input layer detection and fallback dimensions to prevent BFS queue crashes.
* **`deep_learning.py` (Schemas)**: Added the `registry: dict | None` field to the `NodeData` Pydantic model so FastAPI accepts the dynamic contract from the frontend.
* **`layer_registry.json`**: Added `dim_1`, `dim_2`, and `dim_3` parameters to the `input` layer schema so the UI can generate the correct tensor shape fields.

Fixes #177 

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Refactoring

## How Has This Been Tested?

Manually tested the end-to-end pipeline:
- [x] Verified Pydantic accepts the new payload containing the `registry` object.
- [x] Successfully generated a Keras model JSON using the dynamic `getattr()` builder.
- [x] Verified no `KeyError` or `ValueError` crashes occur during the BFS traversal.

## Screenshots

Screenshot 2026-02-28 224929
Screenshot 2026-02-28 224938

### CODE REVIEW

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR successfully delivers an architecturally significant refactor: replacing a brittle hardcoded `if/elif` layer dispatch with a data-driven registry pattern that will scale to new layers without backend code changes. The core security design — server-side registry validation before `getattr()` resolution — is sound and well-reasoned. However, several high-severity bugs undermine correctness and should be resolved before merging: a NullPointerError if `keras_mapping` is absent in the registry, all numeric input fields rendering as `type=\"text\"` due to a wrong type string comparison (`"number"` vs `"int"/"float"`), legacy nodes loaded from the database losing their editable property panels entirely, and Conv2D missing `strides` and `padding` in the new registry — a silent capability regression. Additionally, `generateModelJSON` in `Helpers.jsx` was updated but is now dead code, as `Canvas.jsx` bypasses it with duplicated inline logic.

**Comment by ivantha** [comment]
This branch has merge conflicts with `main`. Please rebase onto the latest `main`, resolve the conflicts, and push an update. Since this stacks on #178 and #180, coordinating the rebase across all three would be cleanest.

**Comment by ivantha** [comment]
Closing this PR as part of a backlog cleanup — it's been inactive for over two weeks with unresolved review feedback or merge conflicts. Thanks for the time you put in.

If you'd still like to land this work, please open a fresh PR rebased on the latest `main`.
